meerk40t/core/undos.py

```diff
@@ -108,3 +108,19 @@ def undolist(self):
         for i, v in enumerate(self._undo_stack):
             q = "*" if i == self._undo_index else " "
             yield f"{q}{str(i).ljust(5)}: state {str(v)}"
+
+    def has_undo(self):
+        with self._lock:
+            if self._undo_index == 0:
+                # At bottom of stack.
+                return False
+            if len(self._undo_stack) == 0:
+                # Stack is entirely empty.
+                return False
+        return True
+
+    def has_redo(self):
+        with self._lock:
```

Comment:
You don't need these locks. If you're not manipulating the data it won't matter if multiple threads look at the same data.